### DISCUSSION TITLE
Wish list for new features?

### DISCUSSION THREAD

**Discussion by tagyoureit** [tagyoureit/nodejs-poolController#610] [General]
We are starting to work on the next version of njsPC and would love to know what you might want to see included!

Here are a few smaller and bigger items I'm tracking (but not promising to deliver!).

* Upgrade to Node v16
* Hardware emulator
* Direct Screenlogic integration
* Some type of user/pin/profile setting on dashPanel
* Generic panels in dashPanel to display non-Pool information

Keeping a more formal list here: https://github.com/users/tagyoureit/projects/1/views/1

**Comment by johnny2678** [comment]
Curious... how can we benefit from direct ScreenLogic integration? 

There are some things about my ET4/Screenlogic shared body setup that make me want to scrap the EasyTouch all together and just go with nixie. But I'm more bark than bite and it takes me forever to implement big changes with all my competing demands for time. 😂

Things like temp setpoints/ modes with schedules, setting Solar deltas, probably some others...

Always exciting to hear the next round of innovation.

**Reply by tagyoureit** [reply]
If you already have njsPC setup with an RS485, it would likely be of little use.  However, for users that may be intimidated to dabble in setting up an RS485, Pi, or think they might void their warranty by hooking a couple of wires into their OCP it would be a lightweight way for them to use njsPC by communicating directly with Screenlogic over ethernet.

**Comment by Telemachus1977** [comment]
Please tackle the wild world of pool color changing lights.

Since none of the companies can seem to come up with a standard selection scheme, it would be great if I could map out my own colors and shows and have NJsPC reflect them easily.

Basically just an option to choose program number ones name and color/scene, then number two and three...  Etc.  Maybe a way for people to share config files with others

**Reply by johnny2678** [reply]
> Please tackle the wild world of pool color changing lights.
> 
Great idea. FL sunseeker LED user here. I never really change my lights anymore because I don't want to have to cycle through a bunch of nodes to find what I want - yes, I know that's the way the light relay works but it's a pain doing it manually.

**Reply by tagyoureit** [reply]
Err, you do know we have this in the app already? 😃  You can thank @rstrouse.  Some of the lights are very strange and it's a bit more than just coming up with a map (at least in a couple instances).

We support:
 - Intellibrite
 - Sunseeker Pooltone
 - Hayward Colorlogic

And it sometimes takes a few iterations, but it's always possible to add new ones.

![image](https://user-images.githubusercontent.com/1646840/190394667-26b3928d-2707-4339-996c-1d7f9a48be05.png)

The mappings can be seen in the code [here](https://github.com/tagyoureit/nodejs-poolController/blob/c0ee451bbfa8eeee92898c76806e7385cacc23ac/controller/boards/NixieBoard.ts#L175)

Additionally, if you want to create CSS colors and gradients we could add those.  The below picture is for the Intellibrite but adding the color overlays for any light is also possible.  See the CSS for Intellibrite [here](https://github.com/rstrouse/nodejs-poolController-dashPanel/blob/d56e1f280c4ddc8b60238ef6806a09653f85ed48/themes/intellibrite.css#L186).

![image](https://user-images.githubusercontent.com/1646840/190395501-b7446479-1925-49b4-873e-3b9c4d391ce5.png)

EDIT: I bet you are going to ask, HOW?
1. Go to circuits and Select Pool Tone as the light type
2. Uncheck Show as Features (which will add it to the Lights section)

![image](https://user-images.githubusercontent.com/1646840/190397416-0bb4b9a2-41e7-416b-8a88-46728b99ff95.png)

And finally, you can create Light Groups if you have 2+ lights.

![image](https://user-images.githubusercontent.com/1646840/190397624-13efc61d-f89b-48a5-8ccd-4fafa3bbf21a.png)

**Reply by Telemachus1977** [reply]
I had no clue, that's very cool!

I assume all of the options are only available using the Nixie controller?  I just pulled the newest version, but I don't see pool tone on my easy touch setup. It is there on my Nixie testing station.

The cheap LED bulbs I'm looking at to replace my 15 year old incandescents (if they ever die) have yet another scheme.  I have been planning on looking at the code to see how difficult it would be to add another scheme.

And I have to live within the restraints of the easy touch system for now.

I guess what I would be asking for in addition to the amazing stuff that is already there, is an onboard ability to map your own lights (limited to the number of options Pentair gives you) and have them stick.

**Comment by Crewski** [comment]
I'd like to see the ability to enable/disable a schedule without it affecting the current status of the equipment.  I know there is a disable button for a pool schedule, but when I click it to disable it turns the pool off.  Say for example I'm going on vacation and don't want my pool lights coming on.  I'd like to just disable that schedule and then re-enable it when I get back.

The other wishlist is the ability to keep the pool on when superchlorinate is happening.  I currently handle stuff like this in Home Assistant but I would love for it to be native to njsPC.

**Reply by tagyoureit** [reply]
Intellicenter has a "vacation" mode.  Maybe disabling schedules could be part of that in Nixie.

EasyTouch (at least) does keep the pool on when super chlorinating.  Will look at that.

**Comment by gw8674** [comment]
@tagyoureit 
Good morning.  On the dashPanel -
The ability to organize the individual schedules, features and light circuits so they are displayed within their boxes on the dashPanel per user preference.
Much like or similar to the way that we can organize the columns (Order of Elements) in the Settings > Appearance page.

Also, any chance that we can break up the "Features" and the "Circuit Groups (macros)" into two distinct boxes on the dashPanel?

Thanks very much!!

**Comment by rock-crusher** [comment]
Hi tagyoureit,

I would like the ability to program events to take place, kinda like a home automation system. I don’t have home automation, not sure how it all works or integrates with Nixie. Maybe someone could help school me on this. 
Anyway something like this:

Clicking a feature button in dash panel “spa night” would turn on landscape lighting, then 4min 30sec later turn on water feature, then 5min 45sec after that fire-up the spa then when the spa reaches temperature the fire ring ignites, then 2min after that the pool/spa lighting turns on, then 3 hours after that everything shuts off. 
Remember, this is just something I pulled out of my *** to use as an example. 
The time delays between each action would be fully modifiable from seconds to hours.

It would also be nice to control sprinklers.

E.G., Click a feature button on dash panel “sprinklers” and station 1 turns on, then after 9min station 1 turns off and station 2 turns on and so on and so forth.

The above events would also be able to be set-up in the schedules section simply by using the event name.  

Since everything in my backyard will be controlled by “Nixie” (pool/spa lights & equipment, yard lighting, patio lights, water feature, fire ring etc.) why not the sprinklers too? Just slap on another relay hat and done.  

Thoughts?

**Reply by Telemachus1977** [reply]
As far as the sprinklers go, what is there to stop you from doing that already?
Throw on another relay hat, name the circuits for each of your sprinkler valves and supply power to them at scheduled times.

Then again, Rachio and other brand sprinkler controllers on the market do this pretty dang well already, with the ability to adjust for weather and rain.

**Reply by rock-crusher** [reply]
Hi Telemachus1977,
Yes, I could schedule each of my sprinkler valves to turn on and off, but that is a pain if you really think about it. Then, when you want to change a schedule or water duration time for an individual valve it affects the entire scheduling, not very practical.   

Agreed, there are plenty of sprinkler controllers on the market and some are actually quite sophisticated. I just thought it would be cool if Nixie could accomplish the same job without having to have additional apps and additional controllers.

**Comment by Telemachus1977** [comment]
I'd love a secure web interface built-in.  Open her up to the world with confidence that Russian hackers won't turn on my heater in the middle of the night.

**Reply by tagyoureit** [reply]
If anyone has Screenlogic, make sure you have a super secure 4 digit passcode.  Lol.  Without it, you could definitely accomplish this act of wastefulness very easily.

**Comment by tagyoureit** [comment]
Add back [basic auth](https://github.com/rstrouse/nodejs-poolController-dashPanel/issues/37).

**Comment by tagyoureit** [comment]
One thing I've been kicking around in my head for a long time is the use of ML and AI.  

Here are some basic things that might be interesting...

* Is my pump running at the same RPMs/watts (given the same conditions) or is something going wrong?
* Can we detect automatically when a pool filter is cleaned?
* Can we determine when a pool filter needs to be cleaned?
* Is my pool equipment in a different situation than I want it to be?  (Eg I want the filter pump off but it's still running)

And if we take it a step further...
* Am I running my pump at the right RPM/flow for the right amount of time to keep the water chemistry balanced?
* Is my pump running at the optimal rpm/flow for the optimal number of hours to save energy?
* Can my schedules self adjust?  If I had a big party over the weekend does the pump need to run more or less?  If the spa was used last night do we need to filter the water more or less?
* Can we estimate how long the pool or spa will take to reach the desired temperature given the current water and air temp, amount of sunlight, whether a cover is installed and knowing how long the pool/spa has taken in the past to reach a desired temperature?

And if we are able to look across many pools in the community (distant future)...
* Is my pump running more or less efficiently than my cohorts given all of the data?
* Which types or pumps are most energy efficient?  Which break down the least?
* Am I using more or less chlorine/acid than others?
* Who's pool is the most energy efficient?  Why?  Can we adjust schedules based on what we learn about others?
* Are different types/doses/strengths of chemicals more or less effective?
* Can we recommend to change out equipment to a different type to save $ or energy?

I've yet to dive into the ML/AI arena but it would be pretty cool to see what supervised/unsupervised learning could accomplish in the pool world.

**Reply by johnny2678** [reply]
+1 for ML/AI

really like the idea of making sure that flow/rpms are adequate to turn the pool over at least 1x / day. Pretty simple for VSF pumps but wondering how this could be done for VS only pumps?

I toyed with automating solar to maximize heat exchange. Every X1 minutes increase the RPMs by X2 and keep repeating every X3 minutes as long as the solar return temp was greater than the solar lead temp by X4 degrees. The thinking is to minimize the time it takes the pool to get to the desired temperature but it comes with a tradeoff of higher average RPMs so I made a Aggressive/Moderate/Conservative mode. Seemed to work pretty well but this was back when my pool logic was in python and I've not moved on and didn't recreate in NodeRed.

I'm currently running my own smart SWG dosing. I gave njspc the old college try but could never get consistent results with the way my pool processes ORP. Would rather see more parameters around dynamic dosing so I could move everything into njspc.

probably could come up with more but your list is a great start!

**Comment by gw8674** [comment]
Deleted

**Comment by michaelmedellin** [comment]
Late to the 8.0.0 party here, but I think it would be super interesting to explore [Matter](https://csa-iot.org/all-solutions/matter/) support for the project. May help streamline the support/maintenance of all the integrations and bindings for different home automation controllers in the wild. 

Interestingly the [Matter 1.0 spec](https://csa-iot.org/wp-content/uploads/2022/11/22-27351-001_Matter-1.0-Device-Library-Specification.pdf#page34) supports pumps and pump controllers as device types.

Maybe not an 8.0.0 effort, but follow on. Considering hacking on a proof of concept to see how viable this would be.

**Reply by tagyoureit** [reply]
Never too late!  Pretty close to releasing v8 but this is just a thread for any future improvements.

Matter could be implemented as an [integration](https://github.com/tagyoureit/nodejs-poolController/wiki/Bindings-Integrations) just like MQTT, Influx, etc.  Or if there is a higher level device it could translate into Matter.  We aren't doing the low level bindings (Zigbee, Z-Wave, Wireless, etc) as we don't want the responsibility of being a Hub ourselves.  Rather, we just want to expose the devices that can then be controlled by the Hub/network.

**Comment by tagyoureit** [comment]
ModBus support for NeoPool - https://github.com/tagyoureit/nodejs-poolController/discussions/559#discussioncomment-6600279

**Comment by CodeSlinger69** [comment]
Something that just came to mind. I use PoolMath from TFP to track my test results and to calculate what chems to add. Would it be useful to have this functionality in njsPC as well? Even in a fully automated Nixie, you still need to manually test certain pool levels, and manually calculate chems to add, such as total alkalinity and calcium hardness.

**Reply by tagyoureit** [reply]
This would be pretty low on the priorities list.  TFP maintains that on a pretty regular basis and we would just be copying what they are already doing well.

**Comment by jamestalmage** [comment]
I'd love to see some more robust rule setting for circuit on/off times.

"Run AT LEAST 4 hours, and until both temperature and ORP setpoints are reached, but don't exceed 10 hours (i.e. something is wrong)"

**Comment by jamestalmage** [comment]
I'd like to see the ability to wire simple green and red buttons to the enclosure, and set rules for them. Red stops whatever program is running, Green starts up a swim program for the next 45 minutes.
